src/platform/audio/alaudiomanager.cpp

```diff
@@ -0,0 +1,1587 @@
+    alListenerfv(AL_ORIENTATION, listenerOri);
+}
+
+/**
+ * Gets the effective volume of the event.s
+ */
+float ALAudioManager::Get_Effective_Volume(AudioEventRTS *event) const
+{
+    float vol = event->Get_Volume() * event->Get_Volume_Shift() * 1.0f;
+
+    // Handle easy cases for none positional stuff.
+    if (event->Get_Event_Info()->Get_Event_Type() == EVENT_MUSIC) {
+        return vol * m_musicVolume;
+    } else if (event->Get_Event_Info()->Get_Event_Type() == EVENT_SPEECH) {
+        return vol * m_speechVolume;
+    } else if (!event->Is_Positional_Audio()) {
+        return vol * m_soundVolume;
+    }
+
+    // Coord3D being used as Vector3 essentially here.
+    Coord3D *event_pos = event->Get_Current_Pos();
+    Coord3D difference = m_listenerPosition;
+    vol *= m_3dSoundVolume;
+
+    if (event_pos == nullptr) {
+        return vol;
+    }
+
+    // Get a vector representing the difference between listerner and sound.
+    difference.Sub(event_pos);
+    float min_dist;
+    float max_dist;
+
+    if (event->Get_Event_Info()->Get_Visibility() & VISIBILITY_GLOBAL) {
+        min_dist = Get_Audio_Settings()->Global_Min_Range();
+        max_dist = Get_Audio_Settings()->Global_Max_Range();
+    } else {
+        min_dist = event->Get_Event_Info()->Min_Range();
+        max_dist = event->Get_Event_Info()->Max_Range();
+    }
+
+    // Calculate distance from our vector between source and listener.
+    float dist = difference.Length();
+
+    // Based on distance and range limits, calculate volume.
+    if (dist > min_dist) {
+        vol *= 1.0f / (float)(dist / min_dist);
+    }
+
+    if (dist >= max_dist) {
+        vol = 0.0f;
+    }
+
+    return vol;
+}
+
+/**
+ * Finds a playing audio that uses a given handle and is a given type.
+ */
+PlayingAudio *ALAudioManager::Find_Playing_Audio_From(uintptr_t handle, unsigned type)
+{
+    switch (type) {
+        case PAT_2DSAMPLE:
+            for (auto it = m_globalAudioList.begin(); it != m_globalAudioList.end(); ++it) {
+                if (*it != nullptr && (*it)->openal.source == (int)handle) {
+                    return *it;
+                }
+            }
+
+            break;
+        case PAT_3DSAMPLE:
+            for (auto it = m_positionalAudioList.begin(); it != m_positionalAudioList.end(); ++it) {
+                if (*it != nullptr && (*it)->openal.source == (int)handle) {
+                    return *it;
+                }
+            }
+
+            break;
+        case PAT_STREAM:
+            for (auto it = m_streamList.begin(); it != m_streamList.end(); ++it) {
+                if (*it != nullptr && (*it)->openal.source == (int)handle) {
+                    return *it;
+                }
+            }
+
+            break;
+        default:
+            break;
+    }
+
+    return nullptr;
+}
+
+/**
+ * Processes the playing audio lists.
+ */
+void ALAudioManager::Process_Playing_List()
```

Comment:
In the original function more stuff is going on in this function. As designed?

Reply:
I have no idea what i did there to be honest :) But i remember that everything except the streams is "fire & forget", so they don't need any processing i guess